# Exercise 4.9 :- Intro to Data Visualization with Python
In this script the following steps are performed
1. Import the new customer Dataframe
2. Do the consistency checks, Data wrangling, address missing values, duplicates, mixed data type checks etc for the customer Dataframe.
3. Combine the cleaned customers Dataframe with the merged order product dataframe. 

# 01 Import the libraries  and the new Dataframe

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# Import the new customers Dataset

In [2]:
path = r'C:\Users\sorna\Desktop\Data Analysis\Achievement 4'

In [3]:
df_customers = pd.read_csv(os.path.join(path, '02.Data', 'Original Data', 'customers.csv'))

# Check the min,max , count values to identify any descripancy in the values.

In [5]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [6]:
#Rename the column as needed
df_customers = df_customers.rename(columns={ 
    'User ID' : 'user_id',
    'Surnam' : 'Surname', 
    'STATE' : 'State', 
    'date_joined' : 'Date_joined', 
    'n_dependants' : 'Number_of_dependents', 
    'fam_status' : 'Family_status', 
    'income' : 'Income'})

In [8]:
#Verify if columns are renamed properly
df_customers.head()

user_id First Name   Surname  Gender       State  Age Date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number_of_dependents Family_status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

# 05 Find the number of missing values in all the columns in the Dataframe

In [9]:
df_customers.isnull().sum()

user_id                     0
First Name              11259
Surname                     0
Gender                      0
State                       0
Age                         0
Date_joined                 0
Number_of_dependents        0
Family_status               0
Income                      0
dtype: int64

In [10]:
#Display the rows with the missing First names
df_first_name = df_customers[df_customers['First Name'].isnull() == True]

In [11]:
df_first_name

user_id First Name   Surname  Gender         State  Age Date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        Number_of_dependents Family_status  Income  
53                         2       married   41709  
73                         0        single   82518  
82                         3       married  117099  
99                         1       married  155673  
105                        3       married  151819  
...                      ...           ...     ...  
206038                     3       married   87783  
206044                     2       married  108488  
206090                     1       married   45275  
206162                     0        single   56325  
206171                     2       married   59222  

[11259 rows x 10 columns]

# There are only missing first name in 11259 records. Hence these records cannot be dropped.

# 05.1 Check for Duplicate records

In [12]:
df_customers = df_customers.drop_duplicates()

In [13]:
df_customers

user_id First Name   Surname  Gender           State  Age Date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        Number_of_dependents Family_status  Income  
0                          3       married  165665  
1                          0        single   59285  
2                          2       married   99568  
3                          0        single   42049  
4                          1       married   40374  
...                      ...           ...     ...  
206204                     1       married  148828  
206205                     3       married  168639  
206206                     2       married   53374  
206207                     1       married   99799  
206208                     1       married   57095  

[206209 rows x 10 columns]

# There are no duplicate records, as the number of records are same after the drop.duplicates function

# 05.2 Check for mixed data types in the Dataframe

In [14]:
for col in df_customers.columns.tolist():
    weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_customers[weird]) > 0:
       print (col)

C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis =

First Name


C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8996\551301605.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis =

In [15]:
df_customers['First Name'].dtype

dtype('O')

#06 Import the latest product_order dataframe and merge with the new cleaned customer dataframe

In [16]:
df_ords_prods_merge1 = pd.read_pickle(os.path.join(path, '02.Data', 'Prepared Data', 'orders_products_flags.pkl'))

In [18]:
df_customers.shape

(206209, 10)

In [19]:
df_ords_prods_merge1.shape

(32404859, 26)

In [20]:
df_merged = pd.merge(df_ords_prods_merge1, df_customers, on = 'user_id', how = 'inner')

In [21]:
df_merged.shape

(32404859, 35)

In [22]:
df_merged.head()

Unnamed: 0  product_id                product_name  aisle_id  \
0           0           1  Chocolate Sandwich Cookies        61   
1           0           1  Chocolate Sandwich Cookies        61   
2           0           1  Chocolate Sandwich Cookies        61   
3           0           1  Chocolate Sandwich Cookies        61   
4           0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  order_id  user_id eval_set  order_number  ...  \
0             19     5.8   3139998      138    prior            28  ...   
1             19     5.8   1977647      138    prior            30  ...   
2             19     5.8    389851      709    prior             2  ...   
3             19     5.8    652770      764    prior             1  ...   
4             19     5.8   1813452      764    prior             3  ...   

     order_freq_flag  First Name  Surname  Gender      State Age Date_joined  \
0  Frequent customer     Charles      Cox    Male  Minnesota  81    8/1/2019   
1  Frequent customer     Charles      Cox    Male  Minnesota  81    8/1/2019   
2  Frequent customer     Deborah    Glass  Female    Vermont  66   6/16/2018   
3  Frequent customer     Heather    Myers  Female  Wisconsin  40    2/9/2020   
4  Frequent customer     Heather    Myers  Female  Wisconsin  40    2/9/2020   

  Number_of_dependents Family_status  Income  
0                    1       married   49620  
1                    1       married   49620  
2                    2       married  158302  
3                    3       married   31308  
4                    3       married   31308  

[5 rows x 35 columns]

08. Export the new merged order, products, customer dataframe to the local storage.

In [23]:
df_merged.to_pickle(os.path.join(path, '02.Data','Prepared Data', 'ords_prods_cust_merge.pkl'))